service/history/statemachine_environment_test.go

```diff
@@ -337,26 +343,52 @@ func TestAccess(t *testing.T) {
 				require.ErrorContains(t, err, "test write error")
 			},
 		},
+		{
+			name:                "write zombie",
+			accessType:          hsm.AccessWrite,
+			workflowState:       enumsspb.WORKFLOW_EXECUTION_STATE_ZOMBIE,
+			expectedSetRequests: 0,
+			accessor: func(n *hsm.Node) error {
+				return fmt.Errorf("accessor should not be called")
+			},
```

Comment:
nit: you could also cast the root node to a mutable state and manipulate it there but this is also fine.

Reply:
hmm the zombie check is at the beginning, and not invoke the accessor if failed, similar to invalid ref.